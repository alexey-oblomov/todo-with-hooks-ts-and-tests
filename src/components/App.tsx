import * as React from 'react';
import {useReducer} from 'react';
import { Action, ActionType, State, ContextState} from '../types/stateType'
import NewTask from "./NewTask";
import TasksList from "./TasksList";

export const initialState: State = {
    newTask: '',
    tasks: []
}

export const ContextApp = React.createContext<Partial<ContextState>>({});

export const todoReducer = (state: State, action: Action):State => {
    switch (action.type) {
        case ActionType.ADD: {
            return {...state, tasks: [...state.tasks, {
                name: action.payload,
                isDone: false
            }]}
        }

        case ActionType.CHANGE: {
            return {...state, newTask: action.payload}
    }
        case ActionType.REMOVE: {
        return {...state, tasks: [...state.tasks.filter(task => task !== action.payload)]}
    }
        case ActionType.TOGGLE: {
            return {...state, tasks: {...state.tasks.map((task) => (task !== action.payload ? task : {...task, isDone: !task.isDone}))}}
        }
        default: throw new Error('Unexpected action');
    }
}


const App: React.FC = () => {
  const [state, changeState] = useReducer<React.Reducer<State, Action>>(todoReducer, initialState);

  const ContextState: ContextState = {
      state,
      changeState
  }

  return (
        <>
            <ContextApp.Provider value={ContextState}>
                <NewTask />
                <TasksList />
            </ContextApp.Provider>
            
        </>
    )
}

export default App;