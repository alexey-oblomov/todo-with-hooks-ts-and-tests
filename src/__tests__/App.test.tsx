import * as React from 'react';
import {shallow} from 'enzyme';
import {fireEvent, render, cleanup} from "@testing-library/react";
import App from "../components/App";

describe('<App />', () => {
    afterEach(cleanup);

    it('Начальный снимок', () => {
        const component = shallow(<App />);
        expect(component).toMatchSnapshot();
    });

    it('Должен рендерить правильное значение input',  async () => {
        const { container } = render(<App/>);
        expect(container.querySelector('input').getAttribute('value')).toEqual('');
        fireEvent.change(container.querySelector('input'), {
            target: {
                value: 'test'
            },
        })
        expect(container.querySelector('input').getAttribute('value')).toEqual('test');
        fireEvent.click(container.querySelector('button'))
        expect(container.querySelector('input').getAttribute('value')).toEqual('');
    });
})