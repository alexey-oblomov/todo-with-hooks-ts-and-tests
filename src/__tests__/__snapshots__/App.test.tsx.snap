// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Начальный снимок 1`] = `
<Fragment>
  <ContextProvider
    value={
      Object {
        "changeState": [Function],
        "state": Object {
          "newTask": "",
          "tasks": Array [],
        },
      }
    }
  >
    <NewTask />
    <TasksList />
  </ContextProvider>
</Fragment>
`;
